README.md
# Meteor Infinite Scroll
**Enables infinite scrolling at the template level**. This package allows you to increment the `limit` parameter of a MongoDB query as the user scrolls down the page. This allows Meteor to use the Oplog Observe Driver for your query, as well as leaving you in control of your publications.

## Usage:

Call `this.infiniteScroll` in the `created` or `rendered` functions for your template.

```js
Template.comments.created = function() {
  // Enable infinite scrolling on this template
  this.infiniteScroll({
    perPage: 20,                        // How many results to load "per page"
    query: {                            // The query to use as the selector in our collection.find() query
        post: 71
    },
    subManager: new SubsManager(),      // (optional, experimental) A meteorhacks:subs-manager to set the subscription on
                                        // Useful when you want the data to persist after this template
                                        // is destroyed.
    collection: 'Comments',             // The name of the collection to use for counting results
    publication: 'CommentsInfinite',     // (optional) The name of the publication to subscribe.
                                        // Defaults to {collection}Infinite
    container: '#selector',              // (optional) Selector to scroll div. Defaults to window
    loadingTemplateName:'loading'       // (optional) Name of loading graphic (spinner) template. Default will show "Loading..."
  });
};
```

Create a publication on the server:

```js
if(Meteor.isServer){
    Meteor.publish('CommentsInfinite', function(limit, query) {
        // Don't use the query object directly in your cursor for security!
        var selector = {};
        check(limit, Number);
        check(query.name, String);
        // Assign safe values to a new object after they have been validated
        selector.name = query.name;

      	return Comments.find(selector, {
          limit: limit,
          // Using sort here is necessary to continue to use the Oplog Observe Driver!
          // https://github.com/meteor/meteor/wiki/Oplog-Observe-Driver
          sort: {
            created: 1
          }
        });
    });
}
```

Render your data as usual. Render the `{{> infiniteScroll }}` template after your data is rendered:

```handlebars
<template name="comments">
    {{#each comments}}
        {{content}}
    {{/each}}
    {{> infiniteScroll }}
</template>
```
> Infinite Scroll will increase the `limit` of the subscription as the `{{> infiniteScroll }}` template approaches the viewport.

Provide data to the template as you usually would. Use `Template.instance().infiniteSub.ready()` like you would use `subscriptionsReady()` on the template instance.
```js
Template.comments.helpers({
  comments: function() {
    return Comments.find({ post: 71 },  {
        limit: Template.instance().infiniteScroll.getLimit(), // optional call to getLimit()
        sort: {
            created: 1
        }
    });
  }
});
```

### Only `limit`

Using `skip` will cause Meteor to use the Polling Observe Driver (see [Oplog Observe Driver in the Meteor Wiki](https://github.com/meteor/meteor/wiki/Oplog-Observe-Driver)). For a full pagination solution that uses skip, check out [alethes:pages](https://github.com/alethes/meteor-pages).

## Styling the loader
The `{{> infiniteScroll }}` template renders:
```html
<template name="infiniteScroll">
  <div class="infinite-load-more">
    <div class="infinite-label">
      Loading...
    </div>
  </div>
</template>
```

When the subscription is loading more data, `.infinite-load-more` will receive the class `loading`. It will be removed when the subscription is marked as ready.

`.infinite-label` is only visible when the subscription is loading.

### API

You can reactively access the current `limit` by using [templateInstance].infiniteScroll.getLimit() (e.g. `Template.instance().infiniteScroll.getLimit()`)

# Todo:
- Customizable threshold for loading more results


CHANGELOG.md
## 1.0.4
- Fix `getLimit()` function (thanks @mizsha)

## 1.0.3
- Add `getLimit()` method to reactively retrieve the current limit, fixes #16

## 1.0.2
- Clear window events when the template is destroyed, fixes #15

## 1.0.1
- Add back loadingTemplateName helper
- Add default loading indicator as a template

## 1.0.0
- Customizable container and loading template (thanks @VolodymyrTymets)
- Removed dburles:mongo-collection-instances dependency (thanks @mizsha)
- Rewritten demo application

## 0.9.2
- Removed `dburles:mongo-collection-instances` dependency, thanks @mizsha

## 0.9.1
- Added `ecmascript` dependency, thanks @JulianKingman

## 0.9.0
- Thanks to @mikepaszkiewicz for his contribution to this release.
- Added example application
- Collection is now looked up by name
- No more ``app`` globals

## 0.8.0
- Remove dependency on Counts, didn't work out so well
- Customizable class on the loader template
- Exposes `InfiniteScroll.triggerLoadMore` so you can use custom scroll regions (eg. `this.$('.results-list').on('resize scroll', _.throttle(InfiniteScroll.triggerLoadMore, 300));`)

## 0.7.2
- Added check package to dependencies

## 0.7.0
- Add sort functionality

## 0.6.1
- Added Tracker.afterFlush when checking to load more data

## 0.6.0
- Overhauled reliability of the loader template and triggers
- Removed `Template.instance().infiniteReady()` in favor of `Template.instance().infiniteSub.ready()`
- New dependency on `tmeasday:publish-counts` for more reliable loading

# 0.5.0
- Fixed an issue with tall window heights

# 0.4.0
- Requires Meteor 1.2 or greater
- Fixed a bug where the loader would not show on the initial subscription load
- Fixed a bug where the subscription query would be mutated by the call to Meteor.subscribe

# 0.3.0
- Minor fixes

# 0.2.0
- Added `subManager` property to set the subscription on a `meteorhacks:subs-manager` Subscription Manager instead of the template. This allows the data to persist when the template is destroyed.

# 0.1.0
- Initial release


package.js
Package.describe({
  name: 'staringatlights:infinite-scroll',
  version: '1.0.4',
  // Brief, one-line summary of the package.
  summary: 'Template level infinite scrolling.',
  // URL to the Git repository containing the source code for this package.
  git: 'https://github.com/abecks/meteor-infinite-scroll',
  // By default, Meteor will default to using README.md for documentation.
  // To avoid submitting documentation, set this field to null.
  documentation: 'README.md'
});

Package.onUse(function(api) {
  api.versionsFrom('1.2');
  api.use(['jquery', 'less', 'reactive-var', 'templating', 'jwo1f:parent-template@0.0.3'], 'client');
  api.use(['check', 'ecmascript'], ['client', 'server']);
  api.export('InfiniteScroll');
  api.addFiles('infiniteScroll.html', 'client');
  api.addFiles('infinite-scroll.js', 'client');
  api.addFiles('infinite-scroll.less', 'client');
});


example/lib/app.js
comments = new Meteor.Collection("comments");
Meteor.startup(function(){
  if(!comments.find().count())
  _(100).times(function(n){
    comments.insert({'post': n, 'content': _.random(1, 9999999) });
   });

});

if(Meteor.isClient){
  Template.body.created = function() {
  // Enable infinite scrolling on this template
  this.infiniteScroll({
    perPage: 10,                        // How many results to load "per page"
    query: {},
                                        // Useful when you want the data to persist after this template
                                        // is destroyed.
    collection: 'comments',             // The name of the collection to use for counting results
    publication: 'CommentsInfinite'     // (optional) The name of the publication to subscribe.
                                        // Defaults to {collection}Infinite
  });
};
  Template.body.helpers({
    comments: function() {
      return comments.find({ },  {
          sort: {
              created: 1
          }
      });
    }
  });
}

if(Meteor.isServer){
    Meteor.publish('CommentsInfinite', function(limit, query) {
        console.log(`${limit} ${JSON.stringify(query)}`);
        // Don't use the query object directly in your cursor for security!
        // var selector = {};
        // check(limit, Number);
        // check(query.name, String);
        // // Assign safe values to a new object after they have been validated
        // selector.name = query.name;

        return comments.find(query, {
          limit: limit,
          sort: {
            created: 1
          }
        });
    });
}


infinite-scroll.js
'use strict';

InfiniteScroll = {};

/**
 * jQuery plugin to determine whether an element is "almost visible".
 * @return {Boolean}
 */
jQuery.fn.isAlmostVisible = function jQueryIsAlmostVisible(container) {
  if (this.length === 0) {
    return;
  }
  var rect = this[0].getBoundingClientRect();

  return (
    rect.top >= 0 &&
    rect.left >= 0 &&
    rect.bottom <= (jQuery(container).height() * 1.5) &&
    rect.right <= jQuery(container).width()
  );
};

/**
 * Enable infinite scrolling on a template.
 */
Blaze.TemplateInstance.prototype.infiniteScroll = function infiniteScroll(options) {
  var tpl = this, _defaults, collection, subManagerCache, limit, subscriber;

  /*
   * Create options from defaults
   */
  _defaults = {
    // How many results to fetch per "page"
    perPage: 10,
    // The query to use when fetching our collection
    query: {},
    // The sorting instructions for MongoDB
    sort: {},
    // The subscription manager to use (optional)
    subManager: null,
    // Collection to use for counting the amount of results
    collection: null,
    // Publication to subscribe to, if null will use {collection}Infinite
    publication: null,
    // Container will use to scroll
    container: window,
    // Custom loading template
    loadingTemplateName: 'infiniteScroll_loading'
  };
  this.infiniteScroll.options = options = _.extend({}, _defaults, options);

  // Validate the options
  check(options.perPage, Number);
  check(options.collection, String);
  check(options.publication, String);
  if (options.container !== window) {
    check(options.container, String);
  }
  check(options.loadingTemplateName, Match.Maybe(String));

  //using meteor connection stores to find collection by name
  let collectionExists = Meteor.connection._stores[options.collection];
  // Collection exists?
  if (!collectionExists) {
    throw new Error('Collection does not exist: ', options.collection);
  } else {
    //set collection to cursor. collectionExists.name evaluates to a string
    collection = collectionExists._getCollection();
  }

  // Generate the publication name if one hasn't been provided
  if (!options.publication) {
    options.publication = options.collection + 'Infinite';
  }

  // If we are using a subscription manager, cache the limit variable with the subscription
  if (options.subManager) {
    // Create the cache object if it doesn't exist
    if (!options.subManager._infinite) {
      options.subManager._infinite = {};
      options.subManager._infinite[options.publication] = {};
    }
    subManagerCache = options.subManager._infinite[options.publication];
  }

  // We use 'limit' so that Meteor can continue to use the OpLogObserve driver
  // See: https://github.com/meteor/meteor/wiki/Oplog-Observe-Driver
  // (There are a few types of queries that still use PollingObserveDriver)
  limit = new ReactiveVar();

  // If the query changes, the limit must reset
  if (options.query instanceof ReactiveVar) {
    tpl.autorun(function () {
      options.query.get();
      limit.set(options.perPage);
    });
  }

  // Retrieve the initial page size
  if (subManagerCache && subManagerCache.limit) {
    limit.set(subManagerCache.limit);
  } else {
    limit.set(options.perPage);
  }

  // Create subscription to the collection
  tpl.autorun(function () {
    // Rerun when the limit changes
    var lmt = limit.get();

    // If a Subscription Manager has been supplied, use that instead to create
    // the subscription. This is useful if you want to keep the subscription
    // loaded for multiple templates.
    if (options.subManager) {
      subscriber = options.subManager;
      // Save the limit in the subscription manager so we can look it up later
      subManagerCache.limit = lmt;
    } else {
      subscriber = tpl;
    }

    var query;
    if (options.query instanceof ReactiveVar) {
      query = options.query.get();
    } else {
      query = options.query;
    }

    var sort;
    if (options.sort instanceof ReactiveVar) {
      sort = options.sort.get();
    } else {
      sort = options.sort;
    }

    tpl.infiniteSub = subscriber.subscribe(options.publication, lmt, query, sort, null);
  });

  /**
   * Load more results if our limit is below the total
   */
  const loadMore = function () {
    var query;
    if (options.query instanceof ReactiveVar) {
      query = options.query.get();
    } else {
      query = options.query;
    }

    var lmt = limit.get(), results = collection.find(query).count();
    if (results >= lmt) {
      limit.set(lmt + options.perPage);
    }
  };

  // Check to see if we need to load more
  const triggerLoadMore = function () {
    if (tpl.$('.infinite-load-more').isAlmostVisible(options.container)) {
      loadMore();
    }
  }

  // Try to load more as soon as soon as the publication is ready,
  // because we might need more documents to fill the screen depending
  // on the perPage setting
  tpl.autorun(function () {
    if (tpl.infiniteSub.ready()) {
      Tracker.afterFlush(triggerLoadMore);
    }
  });

  /**
   * Attempt to trigger infinite loading when resize and scroll browser
   * events are fired.
   */
  $(options.container).on('resize.infiniteScroll scroll.infiniteScroll', _.throttle(triggerLoadMore, 500));

  /**
   * Reactively returns the current query limit.
   */
  this.getLimit = function () {
    return limit.get();
  }
};

Template.infiniteScroll.onDestroyed(function infiniteScrollDestroyed() {
  // Find parent template
  let view = this.view;
  let templateInstance;
  while (view = view.parentView) {
    if (!view) {
      return;
    }

    if (!view.templateInstance) {
      continue;
    }

    var tpl = view.templateInstance();
    if (tpl.infiniteScroll) {
      templateInstance = tpl;
    }
  }
  $(templateInstance.infiniteScroll.options.container).off('resize.infiniteScroll scroll.infiniteScroll');
});


Template.infiniteScroll.helpers({

  loading() {
    // Loop through parent templates until we find infiniteSub
    var tpl = Template.instance();
    while (!tpl.infiniteSub) {
      var parent = tpl.parents();
      if (!parent) {
        break;
      }
      tpl = parent;
    }

    return !tpl.infiniteSub.ready();
  },

  loadingTemplateName() {
    return Template.instance().infiniteScroll.options.loadingTemplateName;
  }
});


demo/demo.js
import { Meteor } from 'meteor/meteor';
import { Mongo } from 'meteor/mongo';
import { Template } from 'meteor/templating';
import { check, Match } from 'meteor/check';

const Comments = new Mongo.Collection("Comments");

if (Meteor.isServer) {
    Meteor.startup(function () {
        if (!Comments.find().count()) {
            _(100).times(function (n) {
                Comments.insert({
                    post: 71,
                    content: _.random(1, 9999999),
                    created: new Date(),
                });
            });
        }
    });

    Meteor.publish('CommentsInfinite', function (limit, query) {
        // Don't use the query object directly in your cursor for security!
        var selector = {};
        check(limit, Number);
        check(query.post, Number);

        // Assign safe values to a new object after they have been validated
        selector.post = query.post;

        return Comments.find(selector, {
            limit: limit,
            // Using sort here is necessary to continue to use the Oplog Observe Driver!
            // https://github.com/meteor/meteor/wiki/Oplog-Observe-Driver
            sort: {
                created: 1
            }
        });
    });
}

if (Meteor.isClient) {
    Template.comments.helpers({
        comments: function () {
            return Comments.find({}, {
                sort: {
                    created: 1
                }
            });
        }
    });

    Template.comments.onCreated(function () {
        // Enable infinite scrolling on this template
        this.infiniteScroll({
            perPage: 20,                        // How many results to load "per page"
            query: {                            // The query to use as the selector in our collection.find() query
                post: 71
            },
            //subManager: new SubsManager(),      // (optional, experimental) A meteorhacks:subs-manager to set the subscription on
            // Useful when you want the data to persist after this template
            // is destroyed.
            collection: 'Comments',             // The name of the collection to use for counting results
            publication: 'CommentsInfinite',    // (optional) The name of the publication to subscribe.
                                                // Defaults to {collection}Infinite
            container: window,                  // (optional) Selector to scroll div.
            // loadingTemplateName: 'loading'   // (optional) Name Of loading template
        });
    });
}